src/write.rs

```diff
@@ -222,15 +222,26 @@ where
                         .unwrap()
                 };
 
+                let free_elements = w_handle.data.capacity() - w_handle.data.len();
+
+                if free_elements < r_handle.data.len() {
+                    // pre-allocate if possible
+                    w_handle.data.reserve(r_handle.data.len() - free_elements);
+                }
+
                 // XXX: it really is too bad that we can't just .clone() the data here and save
                 // ourselves a lot of re-hashing, re-bucketization, etc.
                 w_handle
                     .data
                     .extend(r_handle.data.iter_mut().map(|(k, vs)| {
-                        (
-                            k.clone(),
-                            vs.iter_mut().map(|v| unsafe { v.shallow_copy() }).collect(),
-                        )
+                        (k.clone(), {
+                            // pre-allocate if possible
```

Comment:
I believe `collect()` will already do that if the iterator is `ExactSizeIterator`, which `iter_mut` already provides, so I don't think this will actually have any effect.